pkg/operator/controller/dns_status.go

```diff
@@ -94,32 +176,116 @@ func computeDNSDegradedCondition(oldCondition *operatorv1.OperatorCondition, clu
 			messages = append(messages, "No DNS pods are desired; this could mean all nodes are tainted or unschedulable.")
 		case have == 0:
 			status = operatorv1.ConditionTrue
 			degradedReasons = append(degradedReasons, "NoDNSPodsAvailable")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSNoDNSPodsAvailable, Status: operatorv1.ConditionTrue})
 			messages = append(messages, "No DNS pods are available.")
 		case intstrErr != nil:
 			degradedReasons = append(degradedReasons, "InvalidDNSMaxUnavailable")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSInvalidDNSMaxUnavailable, Status: operatorv1.ConditionUnknown})
 			messages = append(messages, fmt.Sprintf("The DNS daemonset has an invalid MaxUnavailable value: %v", intstrErr))
 		case int(numberUnavailable) > maxUnavailable:
 			status = operatorv1.ConditionTrue
 			degradedReasons = append(degradedReasons, "MaxUnavailableDNSPodsExceeded")
+			degradedConditions = append(degradedConditions, operatorv1.OperatorCondition{Type: DNSMaxUnavailableDNSPodsExceeded, Status: operatorv1.ConditionTrue})
 			messages = append(messages, fmt.Sprintf("Too many DNS pods are unavailable (%d > %d max unavailable).", numberUnavailable, maxUnavailable))
 		}
 	}
 
-	if len(degradedReasons) != 0 {
+	graceConditions, checkedConditions, requeueAfter := checkConditions(expectedConditions, degradedConditions)
+
+	// checkedConditions are degradedConditions without gracePeriods, and the degraded condition is final (this should not be requeued)
+	if len(checkedConditions) != 0 && requeueAfter == 0 {
+		degradedCondition.Status = status
+		degradedCondition.Reason = strings.Join(degradedReasons, "")
+		degradedCondition.Message = strings.Join(messages, "\n")
+		return setDNSLastTransitionTime(degradedCondition, oldCondition), nil
+	}
+
+	if len(degradedReasons) != 0 && requeueAfter == 0 {
 		degradedCondition.Status = status
 		degradedCondition.Reason = strings.Join(degradedReasons, "")
 		degradedCondition.Message = strings.Join(messages, "\n")
 	} else {
+		// graceConditions have grace periods configured in the expectedConditions, so if present, this should be requeued
+		// we check these last, because the other conditions are non-recoverable as degraded conditions
+		if len(graceConditions) != 0 {
+			degradedCondition.Status = operatorv1.ConditionFalse
+			degradedCondition.Reason = strings.Join(degradedReasons, "")
+			degradedCondition.Message = strings.Join(messages, "\n")
+			grace := formatConditions(graceConditions)
+
+			// If the oldCondition changed more recently than requeueAfter period of time, reduce the requeueAfter period of time
+			if oldCondition != nil && (!oldCondition.LastTransitionTime.IsZero()) {
+				t1 := now.Add(-requeueAfter)
+				t2 := oldCondition.LastTransitionTime
+				if t2.After(t1) {
+					requeueAfter = t2.Sub(t1)
+				}
+			}
```

Comment:
I still don't understand the purpose of this code, and I thought you were going to delete it and instead persist all the status conditions so that they would have meaningful, non-zero last transition times (see https://github.com/openshift/cluster-dns-operator/pull/290#discussion_r701569067).

Reply:
I removed the rest of it, but this is needed for when the request has already been requeued, and the condition changed less than 5m ago. In the unit test which has an oldCondition that last transitioned 4 minutes ago, I thought we would want to requeue it in 1 minute rather than wait another 5.  If that's not correct, and we want to requeue in 5 minutes no matter when the last transition occurred, I can change it.

Reply:
> I removed the rest of it, but this is needed for when the request has already been requeued, and the condition changed less than 5m ago. 

We really want to requeue after the remaining portion of the `MaxUnavailableDNSPodsExceeded` condition's grace period.  However, we won't have this condition's last transition time unless we persist it to the API, which is why I thought you were going to change this logic to publish all the status conditions.  

> In the unit test which has an oldCondition that last transitioned 4 minutes ago, I thought we would want to requeue it in 1 minute rather than wait another 5. If that's not correct, and we want to requeue in 5 minutes no matter when the last transition occurred, I can change it.

This part is fine; it's just a question of how we get the last transition time (specifically which condition's last transition time we use).

Reply:
Removed all of it, PTAL when you get a chance @Miciah